### PR TITLE
Remove readonly modifier from EntityManager

### PR DESCRIPTION
**Pull request by nicolas-grekas**
The readonly modifier has been added as if it had no consequences, but in reality it breaks the possibility to reset the EM, while this is an important capability when using it in long-running processes.

This has the same reasoning as why `@final` is used and not the real final modifier.

Please see the discussion in https://github.com/symfony/symfony/issues/54228 for more detailed explanations.

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
I added a test case that breaks if the unitOfWork property is made readonly or if the class is made final.

**Review by derrabus — DISMISSED** [review]

**Comment by derrabus** [comment]
> The readonly modifier has been added as if it had no consequences, but in reality it breaks the possibility to reset the EM, while this is an important capability when using it in long-running processes.

FTR: Resetting the EM is a feature that the ORM neither advertised nor documented. Reopening an EM is explicitly not supported. I'm fine merging this change to unblock downstream projects. But this reset feature should be implemented properly in this repository. We cannot maintain or support features that are monkey-patched onto our code.

**Comment by greg0ire** [comment]
I will review this after I find time to read https://github.com/doctrine/orm/issues/5933

File: tests/Tests/EntityManagerTest.php
Comment (greg0ire) [inline tests/Tests/EntityManagerTest.php]: https://github.com/doctrine/orm/issues/5933 should probably be mentioned here, in one form or another.

File: src/EntityManager.php
Comment (greg0ire) [inline src/EntityManager.php:1]: Does each and every property in this diff really need to be mutable? For instance, `Configuration` does not have a reference to entity manager, so I don't see why this is necessary. Maybe because of how `resetLazyObject()` works?

**Comment by greg0ire** [comment]
This does not fix a bug with Doctrine, and as such should target 3.2.x. We may release 3.2.0 quickly after this is merged as a courtesy to Symfony users.

File: src/EntityManager.php
Comment (nicolas-grekas) [reply src/EntityManager.php:1]: Yes, because resetting a lazy object means resetting all its properties, and calling its full instantiation logic again on-demand.

File: tests/Tests/EntityManagerTest.php
Comment (nicolas-grekas) [reply tests/Tests/EntityManagerTest.php]: I've added a comment about this

**Comment by nicolas-grekas** [comment]
> This does not fix a bug with Doctrine, and as such should target 3.2.x. We may release 3.2.0 quickly after this is merged as a courtesy to Symfony users.

Fine to me :rocket: :shipit: :ship:

**Review by greg0ire — APPROVED** [review]

**Comment by curry684** [comment]
> We may release 3.2.0 quickly after this is merged as a courtesy to Symfony users.

My mailbox highly appreciates this   😉 

image

**Review by konopskyyy — APPROVED** [review]

**Comment by stof** [comment]
@derrabus the feature request to move this feature to the ORM itself exists since 2016: https://github.com/doctrine/orm/issues/5933
Being able to reopen an EntityManager instance (after clearing it to remove the broken state of managed entities) is the only way you can use the ORM in a long running queue consumer as you don't want to kill your long-running process each time an exception happen during the handling of a message.

**Comment by derrabus** [comment]
I think, we both know the difference beware _requesting_ a feature and _implementing_ it. I'm aware that the request exists, bit somebody needs to do it.

**Pull request merged**
